compiler/src/iree/compiler/Dialect/Flow/TransformExtensions/FlowExtensions.cpp

```diff
@@ -446,5 +447,286 @@ transform_dialect::ForeachThreadToFlowDispatchWorkgroupsOp::applyToOne(
+                                regionOp.getResultTypes().end());
+  resultTypes.push_back(result.getType());
+
+  // Create new DispatchRegionOp and move over the body.
+  auto newRegionOp = rewriter.create<Flow::DispatchRegionOp>(
+      regionOp->getLoc(), resultTypes, dynamicDims);
+  newRegionOp.getBody().takeBody(regionOp.getBody());
+  rewriter.replaceOp(
+      regionOp, newRegionOp.getResults().take_front(regionOp->getNumResults()));
+
+  // Update terminator.
+  Flow::ReturnOp returnOp =
+      cast<Flow::ReturnOp>(newRegionOp.getBody().front().getTerminator());
+  SmallVector<Value> returnedValues(returnOp.getOperands().begin(),
+                                    returnOp.getOperands().end());
+  returnedValues.push_back(result);
+  returnOp.operandsMutable().assign(returnedValues);
+
+  return newRegionOp;
+}
+
+/// Return `true` if the given op post-dominates the dispatch region.
+static bool isAfterRegion(Operation *op, Flow::DispatchRegionOp regionOp) {
+  Operation *ancestor = regionOp->getBlock()->findAncestorOpInBlock(*op);
+  assert(ancestor && "expected that op and regionOp are in the same block");
+  return regionOp->isBeforeInBlock(ancestor);
+}
+
+// Clone a `target` op that is preceding the given dispatch region op into the
+// dispatch region.
+//
+// All uses of the target inside of the dispatch region are replaced with the
+// results of the cloned op.
+//
+// If `updateUsesOutsideOfRegion` is set, all uses of the target op after the
+// dispatch region, are also updated: The target op's results are returned from
+// the dispatch region an used in those places.
+//
+// Example when `updateUsesOutsideOfRegion` is unset:
+//
+// %0 = "some_op"() : () -> (tensor<?xf32>)
+// %r = flow.dispatch.region -> (tensor<?xf32>{%d0}) {
+//   %1 = "another_op"(%0) : (tensor<?xf32>) -> (tensor<?xf32>)
+//   flow.return %1 : tensor<?xf32>
+// }
+// %2 = "yet_another_use"(%0) : (tensor<?xf32>) -> (tensor<?xf32>)
+//
+// In this example, "some_op" will be cloned into the dispatch region and the
+// OpOperand of "another_op" will be replaced:
+//
+// %0 = "some_op"() : () -> (tensor<?xf32>)
+// %r = flow.dispatch.region -> (tensor<?xf32>{%d0}) {
+//   %0_clone = "some_op"() : () -> (tensor<?xf32>)
+//   %1 = "another_op"(%0_clone) : (tensor<?xf32>) -> (tensor<?xf32>)
+//   flow.return %1 : tensor<?xf32>
+// }
+// %2 = "yet_another_use"(%0) : (tensor<?xf32>) -> (tensor<?xf32>)
+static FailureOr<Flow::DispatchRegionOp> clonePrecedingOpIntoDispatchRegion(
+    RewriterBase &rewriter, Operation *target, Flow::DispatchRegionOp regionOp,
+    bool updateUsesOutsideOfRegion) {
+  assert(target->isBeforeInBlock(regionOp) &&
+         "expected that target comes first");
+  Block &body = regionOp.getBody().front();
+
+  // Gather all uses of `target`.
+  SmallVector<OpOperand *> usesInsideOfRegion, usesOutsideOfRegion;
+  bool hasUsesBeforeRegion = false;
+  for (OpOperand &use : target->getUses()) {
```

Comment:
I am still not clear why you are looking into uses of the target.... I know I said dont use dominator, but this is actually using the same methods that the dominator uses. Look at the implementation of `isBeforeInBlock` [here](https://github.com/llvm/llvm-project/blob/c1f65df19c0f7324db8a7ec235308aee55fe22ef/mlir/lib/IR/Operation.cpp#L261) . The `recomputeOpOrder` is what is the problem. So this still has the same issue.

Reply:
Note the `isBeforeInBlock` is only used in the `assert`. It's just there to catch incorrect usage API usage, which can result in dominance errors. These can be difficult to debug. Dominance errors may surface later (sometimes it's not clear which transform caused them) and sometimes they are accidentally fixed up by canonicalizers, CSE, etc. Meaning that the dominance error may not surface at all during our tests and when we run another model/benchmark it may suddenly generate incorrect IR.

Reply:
I'm checking the uses so that I know all places that must be updated. This should be cheap.